docs/languages/csharp-iterators.md

```diff
@@ -0,0 +1,247 @@
+# Introduction
+
+Almost every program you write will have some need to iterate
+over a collection. You'll write code that examines every item in
+a collection. You'll also create classes
+that produce a collection by creating a method that produces an
+iterator for the elements of that class. The C# language provides
+features for both these scenarios. This article provides an overview
+of those features.
+
+# iterating with foreach
+
+Enumerating a collection is simple: The `foreach` keyword enumerates
+a collection, executing the embedded statement once for each element
+in the collection:
+ 
+```cs
+foreach (var item in collection)
+{
+   Console.WriteLine(item.ToString());
+}
+```
+
+That's all there is to it. To iterate over all the contents of a collection,
+the `foreach` statement is all you need. The `foreach` statement isn't magic,
+though. It relies on two interfaces defined in the .NET core library in order
```

Comment:
Perhaps a link or brief explanation about how `IEnumerable<T>` and `IEnumerator<T>` tie into this?  I like calling them out here, but without any info (however basic) about why these interfaces are important for iterating I would end up slightly confused.